### ISSUE TITLE
endwin() from signal handler?

### ISSUE THREAD

**Issue by tj90241** [philipdexter/vx#1]
Is endwin() signal safe? It talks about flushing buffers and such [EDIT: in the man pages], so it shouldn't be, right...?

I would suggest to set lets_edit to false, then use siglongjmp/sigsetjmp to exceptionally jump back right before the main while loop, so that endwin() does it's work outside of a signal handling context... but that's POSIX (not C99), so I'm not sure you want to introduce that solution.

**Comment by tj90241** [comment]
FWIW, exit() isn't signal safe either.

See: http://stackoverflow.com/questions/8833394/can-exit-fail-to-terminate-process

**Comment by philipdexter** [comment]
I highly doubt it's signal safe. I figured I was doing some bad stuff in that handler but have ignored it for a while.

I'd be fine with the longjmp solution. I have no opinion on POSIX vs C99. I suppose if it's POSIX that means it probably wouldn't work on windows, correct? I don't have a problem with that (at least yet...)

About exit(), ah yes I think I remember that. Is it safe to replace `exit` with `_exit`? Some quick research seems to say yes. Although if we change the behavior as mentioned above I don't think we need `exit` in the signal handler anyways.

**Comment by philipdexter** [comment]
I used setjmp/longjmp instead of the sig counterparts because they don't exist on my system. Are they widespread? `sigjmp` saves the signal mask on my system by default.

**Comment by tj90241** [comment]
Yep, _exit is safe.

siglongjmp and sigsetjmp should exist... did you define _POSIX_C_SOURCE? I'm fairly certain you cannot longjmp out of a signal handler reliably - it is undefined as to whether or not the signal mask is saved by the non-sig counterparts.

Signal handling is hard.

**Comment by philipdexter** [comment]
Ah, I was missing the _POSIX_C_SOURCE.

**Comment by tj90241** [comment]
Now I'm kind of doubting myself as to whether you can even siglongjmp out of a signal handler. :(

See list of functions: http://man7.org/linux/man-pages/man7/signal.7.html

**Comment by philipdexter** [comment]
Hm, I'm not sure what to do then. Any suggestions? How do other projects
handle this?

**Comment by tj90241** [comment]
I don't know. It might be the case that you can, it might not. Lots of quibbering on stackoverflow and the likes. I need to crack open TLPI or something.

Usually I just set a flag in the signal handler, and then rely on the main loop to break out.

**Comment by tj90241** [comment]
http://www.gnu.org/software/libc/manual/html_node/Longjmp-in-Handler.html

Conclusion: calling the function itself is not undefined, BUT you can still experience inconsistent data in the even that you're not masking signals around the update to avoid asynchronous events coming in and ruining your day.

tl;dr: it won't work in the way that we need it to.

Most applications only use a signal handler to force blocking calls to break out with EINTR (i.e., recv and send) and set a global flag to indicate that it's time to bail. I don't think any amount of magic will make global jumps work in the way that we'd like them to for the handler...

**Comment by philipdexter** [comment]
Alright, I guess we'll just change the lets_edit flag
